File: day10/src/main/python/solution.py
def read_input():
    f = open('../resources/input', 'r')
    # f = open('../../test/resources/part1/example1.input', 'r')
    return list(map(lambda l: parse_line(l.strip()), f.readlines()))


opening_chars = {'(', '<', '{', '['}


def closing_char(c):
    if c == '(':
        return ')'
    elif c == '<':
        return '>'
    elif c == '{':
        return '}'
    else:
        return ']'


def parse_line(line):
    invalid = ' '
    stack = []
    for c in line:
        if c in opening_chars:
            stack.append(c)
        elif closing_char(stack[-1]) == c:
            stack.pop()
        else:
            invalid = c
            break
    return invalid, stack


def score_compiler_error(errors):
    if errors[0] == ' ':
        return 0
    elif errors[0] == ')':
        return 3
    elif errors[0] == ']':
        return 57
    elif errors[0] == '}':
        return 1197
    else:
        return 25137


def score_compiler_errors(lines):
    return sum(map(score_compiler_error, lines))


def score_finish_error(errors):
    if errors[0] != ' ':
        return 0
    count = 0
    for c in reversed(errors[1]):
        count *= 5
        if closing_char(c) == ')':
            count += 1
        elif closing_char(c) == ']':
            count += 2
        elif closing_char(c) == '}':
            count += 3
        else:
            count += 4
    return count


def score_incomplete_errors(lines):
    scores = sorted(list(filter(lambda x: x != 0, map(score_finish_error, lines))))
    return scores[len(scores) // 2]


if __name__ == "__main__":
    instructions = read_input()
    print("part 1")
    print(score_compiler_errors(instructions))
    print("part 2")
    print(score_incomplete_errors(instructions))

File: day10/src/main/cpp/solution.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <set>
#include <stack>
#include <string>
#include <vector>

using namespace std;

#define compilerResult pair<char, stack<char>>

const set<char> openingChars = {'(', '[', '<', '{'};

char closingChar(char c) {
  switch (c) {
  case '(':
    return ')';
  case '[':
    return ']';
  case '<':
    return '>';
  case '{':
    return '}';
  default:
    throw 1;
  }
}

compilerResult compileLine(string line) {
  char invalid = ' ';
  stack<char> openend;
  for (auto const &c : line) {
    if (openingChars.find(c) != openingChars.end()) {
      openend.push(c);
    } else if (closingChar(openend.top()) == c) {
      openend.pop();
    } else {
      invalid = c;
      break;
    }
  }
  return pair(invalid, openend);
}

vector<compilerResult> readInput() {
  ifstream file("../resources/input");
  vector<compilerResult> result;
  string s;
  while (getline(file, s)) {
    result.push_back(compileLine(s));
  }
  return result;
}

long long calculateSyntaxErrorScore(vector<compilerResult> problems) {
  long long count = 0;
  for (auto const &problem : problems) {
    switch (problem.first) {
    case ')':
      count += 3;
      break;
    case ']':
      count += 57;
      break;
    case '}':
      count += 1197;
      break;
    case '>':
      count += 25137;
      break;
    default:
      continue;
    }
  }
  return count;
}

long long calculateMissingScore(vector<compilerResult> problems) {
  vector<long long> score;
  for (auto const &problem : problems) {
    if (problem.first != ' ') {
      continue;
    }
    stack<char> charStack = problem.second;
    long long count = 0;
    while (!charStack.empty()) {
      count *= 5;
      switch (charStack.top()) {
      case '(':
        count += 1;
        break;
      case '[':
        count += 2;
        break;
      case '{':
        count += 3;
        break;
      case '<':
        count += 4;
        break;
      default:
        throw 2;
      }
      charStack.pop();
    }
    score.push_back(count);
    
  }
  sort(score.begin(), score.end());
  return score[score.size() / 2];
}

int main() {
  vector<compilerResult> problems = readInput();
  cout << "part 1" << endl
       << calculateSyntaxErrorScore(problems) << endl
       << "part 2" << endl
       << calculateMissingScore(problems) << endl;
  return 0;
}